crates/fuel-core/src/state/rocks_db_key_iterator.rs

```diff
@@ -0,0 +1,81 @@
+use crate::database::{
+    Error as DatabaseError,
+    Result as DatabaseResult,
+};
+use anyhow::anyhow;
+
+use fuel_core_storage::iter::IterDirection;
+use rocksdb::{
+    DBAccess,
+    DBRawIteratorWithThreadMode,
+};
+
+pub enum OwnedIteratorMode {
+    Start,
+    End,
+    From(Vec<u8>, rocksdb::Direction),
+}
+
+pub struct RocksDBKeyIterator<'a, D: DBAccess> {
+    raw: DBRawIteratorWithThreadMode<'a, D>,
+    direction: IterDirection,
+    done: bool,
+}
+
+impl<'a, D: DBAccess> Iterator for RocksDBKeyIterator<'a, D> {
+    // decoupling the Key type from crate::storage::KeyItem
+    type Item = DatabaseResult<Vec<u8>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            None
+        } else if let Some(key) = self.raw.key() {
+            let key = key.to_vec();
```

Comment:
I'm curious. Is it possible to return a slice instead of the vector? In most cases we decode key and don't use the vector itself

Reply:
not without adding additional lifetime constraints :/

Reply:
Okay, we can try to play with it in follow up PR=)

Reply:
https://github.com/FuelLabs/fuel-core/issues/2097#issue-2471520476